### PR TITLE
Make use of Path/Path.IO as much as possible

### PR DESCRIPTION
**Pull request by jcmincke**
### Pull Request Description

Make use of Path/Path.IO as much as possible. See Issue:https://github.com/luna/luna/issues/404

### Important Notes

Please, look at the TODO [JCM]. These are open questions/proposals.

I created a new package: luna-path which contains functions not available in the path package.
 
### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] The code conforms to the [Haskell](https://github.com/luna/luna/blob/master/doc/haskell-style-guide.md) or [Scala](https://github.com/luna/luna/blob/master/doc/scala-style-guide.md) style guides as appropriate.
- [x] The code has been tested where possible.

### CODE REVIEW

File: shell/src/Luna/Shell/Command.hs [hunk 1]
```diff
@@ -170,23 +169,25 @@ data CommandOpts
...
 
             if fileExists then do
-                filePath <- Path.parseAbsFile canonicalPath
-                if Path.fileExtension filePath /= Package.lunaFileExt then
-                    hPutStrLn stderr $ canonicalPath <> " is not a Luna file."
-                else Interpret.file filePath stdlibPath
-            else if projectExists then runPackage canonicalPath stdlibPath
-            else hPutStrLn stderr $ target <> " not found."
+                if Path.fileExtension canonicalFilePath /= Package.lunaFileExt then
+                    hPutStrLn stderr $ (Path.fromAbsFile canonicalFilePath) <> " is not a Luna file."
```
Comment (iamrecursion) [inline shell/src/Luna/Shell/Command.hs]: Overly long line.

File: shell/src/Luna/Shell/Command.hs [hunk 2]
```diff
@@ -338,25 +346,29 @@ acquireGlobalConfig = liftIO $ Exception.catch acquire recovery where
...
-        let lunaConfigFile = lunaConfigDir </> Global.configName
-        configFileExists <- Directory.doesFileExist lunaConfigFile
+        let lunaConfigFile = lunaConfigDir </> configNameRelFilePath
+        configFileExists <- Path.doesFileExist lunaConfigFile
 
         unless configFileExists $ do
             putStrLn "Generating Global Config"
-            putStrLn $ "Please fill in your name in " <> lunaConfigFile
-            putStrLn $ "Please fill in your email in " <> lunaConfigFile
+            putStrLn $ "Please fill in your name in " <> (Path.fromAbsFile lunaConfigFile)
```
Comment (iamrecursion) [inline shell/src/Luna/Shell/Command.hs]: I'd replace the duplication here with a `let fpPath = Path.fromAbsFile lunaConfigFile` for cleanliness' sake.

File: shell/src/Luna/Shell/GenerateDocumentation.hs [hunk 3]
```diff
@@ -75,13 +75,14 @@ documentUnit name unit = UnitDocumentation name Nothing classDocs funDocs where
 defaultOutFileName :: FilePath
 defaultOutFileName = "documentation.json"
 
+
```
Comment (iamrecursion) [inline shell/src/Luna/Shell/GenerateDocumentation.hs]: Extraneous whitespace.

File: stack.yaml [hunk 4]
```diff
@@ -74,3 +75,6 @@ packages:
     - lib/vector-text
     - lib/yaml-utils
 
+extra-deps: [
```
Comment (iamrecursion) [inline stack.yaml]: Extra deps don't live here. This should be added to the list in `config/snapshot.yaml` instead.

File: stdlib/src/Luna/Prim/DynamicLinker.hs [hunk 5]
```diff
@@ -69,29 +71,29 @@ instance Exception NativeLibraryLoadingException where
...
     projectLocalDirs      <- findLocalNativeLibsDirs projectDir
     pure $ projectNativeDirs : projectLocalDirs
 
-tryLoad :: FilePath -> IO (Either String Handle)
+tryLoad :: Path Abs File -> IO (Either String Handle)
 tryLoad path = do
     loadRes <- tryAny $ nativeLoadLibrary path
     let errorDetails exc =
-            "loading \"" <> path <> "\" failed with: " <> displayException exc
+            "loading \"" <> (Path.fromAbsFile path) <> "\" failed with: " <> displayException exc
```
Comment (iamrecursion) [inline stdlib/src/Luna/Prim/DynamicLinker.hs]: Line length!

File: path/src/Luna/Path.hs [hunk 6]
```diff
@@ -0,0 +1,3 @@
+module Luna.Path (module X) where
```
[the remark this answers is not in the document]
Comment (jcmincke) [reply path/src/Luna/Path.hs]: I thought that having one module per package which re-exports everything was the norm.
Fine with me, I'll import from Luna.Path.Path

File: shell/src/Luna/Shell/CWD.hs [hunk 7]
```diff
@@ -22,9 +21,9 @@ appImageEnvVar = "OWD"
...
-get :: MonadIO m => m FilePath
+get :: MonadIO m => m (Path Abs Dir)
 get = do
-    cwd <- liftIO . catch (Environment.getEnv appImageEnvVar)
-            $ \(_ :: SomeException) -> liftIO Directory.getCurrentDirectory
-    liftIO $ Directory.canonicalizePath cwd
-
+    cwd <- liftIO . catch (Environment.getEnv appImageEnvVar >>= parseAbsDir)
+            $ \(_ :: SomeException) -> liftIO getCurrentDir
+    path <- liftIO $ canonicalizePath cwd
```
[the remark this answers is not in the document]
Comment (jcmincke) [reply shell/src/Luna/Shell/CWD.hs]: You're right, liftIO is not even needed.

File: path/src/Luna/Path/Path.hs [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+-- @dirnameNoSlash $(mkRelDir ".") == $(mkRelDir ".")@
+--
+-- @dirnameNoSlash (p <\> a) == dirnameNoSlash a@
+--
+dirnameNoSlash :: Path b Dir -> Path Rel Dir
+dirnameNoSlash (IPath.Path "") = IPath.Path ""
+dirnameNoSlash (IPath.Path l) | FilePath.isDrive l = IPath.Path ""
+dirnameNoSlash (IPath.Path l) = IPath.Path (last (FilePath.splitDirectories l))
+
+dropExtensions :: Path a b -> Path a b
```
[the remark this answers is not in the document]
Comment (jcmincke) [reply path/src/Luna/Path/Path.hs]: Done. But I'd like to add some test around these. Anyway, I can add them later.

**Comment by jcmincke** [comment]
Thanks for your review @iamrecursion.

File: path/package.yaml [hunk 9]
```diff
@@ -0,0 +1,22 @@
+_config/lib: !include "../config/hpack-common.yaml"
+
+name:       luna-path
+version:    "0.1"
+author:     Jean-Christophe Mincke <jeanchristophe.mincke@gmail.com>
+maintainer:
+    - Jean-Christophe Mincke <jeanchristophe.mincke@gmail.com>
```
Comment (iamrecursion) [inline path/package.yaml]:
```suggestion
    - Luna Team <contact@luna-lang.org>
    - Jean-Christophe Mincke <jeanchristophe.mincke@gmail.com>
```

We just need to make it clear that you don't bear the sole responsibility for maintaining this!

File: stdlib/src/Luna/Prim/DynamicLinker.hs [hunk 10]
```diff
@@ -163,58 +169,68 @@ nativeLoadSymbol handle symbol = Foreign.castPtrToFunPtr
 dynamicLibraryExtensions :: [String]
 dynamicLibraryExtensions = ["", ".dll"]
 
-nativeLibraryProjectDir :: String
-nativeLibraryProjectDir = "windows"
+nativeLibraryProjectDir :: Path Rel Dir
+nativeLibraryProjectDir = $(mkRelDir "windows")
```
Comment (iamrecursion) [inline stdlib/src/Luna/Prim/DynamicLinker.hs]:
```suggestion
nativeLibraryProjectDir = $(Path.mkRelDir "windows")
```

**Comment by iamrecursion** [comment]
Thank you so much Jean-Christophe! This is a brilliant contribution, and hopefully the first of many!

**Pull request merged**